beginner_source/text_sentiment_ngrams_tutorial.py

```diff
@@ -156,188 +186,211 @@ def generate_batch(batch):
 #
 
 
-######################################################################
-# `torch.utils.data.DataLoader <https://pytorch.org/docs/stable/data.html?highlight=dataloader#torch.utils.data.DataLoader>`__
-# is recommended for PyTorch users, and it makes data loading in parallel
-# easily (a tutorial is
-# `here <https://pytorch.org/tutorials/beginner/data_loading_tutorial.html>`__).
-# We use ``DataLoader`` here to load AG_NEWS datasets and send it to the
-# model for training/validation.
-#
-
-from torch.utils.data import DataLoader
+import time
 
-def train_func(sub_train_):
+def train(model, dataloader):
+    model.train()
+    total_acc, total_count = 0, 0
+    log_interval = 500
+    start_time = time.time()
 
-    # Train the model
-    train_loss = 0
-    train_acc = 0
-    data = DataLoader(sub_train_, batch_size=BATCH_SIZE, shuffle=True,
-                      collate_fn=generate_batch)
-    for i, (text, offsets, cls) in enumerate(data):
+    for idx, (label, text, offsets) in enumerate(dataloader):
         optimizer.zero_grad()
-        text, offsets, cls = text.to(device), offsets.to(device), cls.to(device)
-        output = model(text, offsets)
-        loss = criterion(output, cls)
-        train_loss += loss.item()
+        predited_label = model(text, offsets)
+        loss = criterion(predited_label, label)
         loss.backward()
+        torch.nn.utils.clip_grad_norm_(model.parameters(), 0.1)
         optimizer.step()
-        train_acc += (output.argmax(1) == cls).sum().item()
-
-    # Adjust the learning rate
-    scheduler.step()
+        total_acc += (predited_label.argmax(1) == label).sum().item()
+        total_count += label.size(0)
+        if idx % log_interval == 0 and idx > 0:
+            elapsed = time.time() - start_time
+            print('| epoch {:3d} | {:5d}/{:5d} batches '
+                  '| accuracy {:8.3f}'.format(epoch, idx, len(dataloader),
+                                              total_acc/total_count))
+            total_acc, total_count = 0, 0
+            start_time = time.time()
```

Comment:
unused variable?

Reply:
Here we reset the start_time variable to have the new time period.